File: backend/school.py
"""Bounded, household-owned school routines; never provider credentials.

Civil dates are deliberate: school weeks do not change with DST. Term ends are
inclusive. Week A continues through holidays unless a term supplies a new anchor.
"""
from __future__ import annotations
from datetime import date
from typing import Literal
from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator
import json
import re

Uniform = Literal['school', 'pe', 'none', 'unknown']


def civil(value: str) -> str:
    if not re.fullmatch(r'\d{4}-\d{2}-\d{2}', value):
        raise ValueError('Use a date in YYYY-MM-DD format.')
    parsed = date.fromisoformat(value)
    if not 1900 <= parsed.year <= 2200:
        raise ValueError('School dates must be between 1900 and 2200.')
    return value


def monday(value: str) -> str:
    civil(value)
    if date.fromisoformat(value).weekday() != 0:
        raise ValueError('Week A must be anchored to a Monday.')
    return value


class SchoolModel(BaseModel):
    model_config = ConfigDict(extra='forbid', strict=True)

    @field_validator('*', mode='before')
    @classmethod
    def text_controls(cls, value):
        if isinstance(value, str) and re.search(r'[\x00-\x1f\x7f]', value):
            raise ValueError('Control characters are not allowed.')
        return value


class Term(SchoolModel):
    start: str
    end: str
    weekA: str = ''

    @model_validator(mode='after')
    def valid(self):
        civil(self.start)
        civil(self.end)
        if self.end < self.start:
            raise ValueError('Term end must not precede its start.')
        if self.weekA:
            monday(self.weekA)
        return self


class Lesson(SchoolModel):
    name: str = Field(min_length=1, max_length=80)
    start: str = ''
    end: str = ''
    room: str = Field(default='', max_length=40)

    @model_validator(mode='after')
    def valid(self):
        self.name = self.name.strip()
        if not self.name:
            raise ValueError('A lesson or activity needs a name.')
        if bool(self.start) != bool(self.end):
            raise ValueError('Enter both start and end times, or leave both blank.')
        for value in (self.start, self.end):
            if value and not re.fullmatch(r'(?:[01]\d|2[0-3]):[0-5]\d', value):
                raise ValueError('Use a time in HH:MM format.')
        if self.start and self.end <= self.start:
            raise ValueError('End time must be later on the same day.')
        return self


class SchoolDay(SchoolModel):
    week: int = Field(ge=1, le=2)
    weekday: int = Field(ge=1, le=5)  # ISO Monday to Friday
    uniform: Uniform = 'unknown'
    snacks: int | None = Field(default=None, ge=0, le=9)
    kit: str = Field(default='', max_length=120)
    lessons: list[Lesson] = Field(default_factory=list, max_length=10)


class Activity(Lesson):
    weekday: int = Field(ge=1, le=7)
    week: int = Field(default=0, ge=0, le=2)  # 0 = every week
    notes: str = Field(default='', max_length=160)
    skipDates: list[str] = Field(default_factory=list, max_length=80)

    @model_validator(mode='after')
    def activity_valid(self):
        if not self.start:
            raise ValueError('Clubs and music lessons need start and end times.')
        for value in self.skipDates:
            civil(value)
        if len(set(self.skipDates)) != len(self.skipDates):
            raise ValueError('Cancelled dates must be unique.')
        return self


class SchoolProfile(SchoolModel):
    member: str = Field(min_length=1, max_length=48, pattern=r'^[a-zA-Z0-9_-]+$')
    cycleWeeks: int = Field(default=1, ge=1, le=2)
    anchor: str
    terms: list[Term] = Field(default_factory=list, max_length=32)
    excludeDates: list[str] = Field(default_factory=list, max_length=180)
    days: list[SchoolDay] = Field(default_factory=list, max_length=10)
    activities: list[Activity] = Field(default_factory=list, max_length=24)

    @model_validator(mode='after')
    def valid(self):
        monday(self.anchor)
        terms = sorted(self.terms, key=lambda t: t.start)
        if any(a.end >= b.start for a, b in zip(terms, terms[1:])):
            raise ValueError('Term ranges must not overlap; both end dates are inclusive.')
        for value in self.excludeDates:
            civil(value)
        if len(set(self.excludeDates)) != len(self.excludeDates):
            raise ValueError('Excluded dates must be unique.')
        keys = [(d.week, d.weekday) for d in self.days]
        if len(set(keys)) != len(keys):
            raise ValueError('Only one school-day rule per weekday and cycle week.')
        if any(d.week > self.cycleWeeks for d in self.days) or any(a.week > self.cycleWeeks for a in self.activities):
            raise ValueError('Week B rules require a two-week timetable.')
        return self


def validate_school_members(profiles: list[SchoolProfile], members) -> None:
    keys = [p.member for p in profiles]
    if len(keys) != len(set(keys)) or not set(keys) <= {m.key for m in members}:
        raise ValueError('School profiles must use distinct, configured people.')
    # Stay below the existing 64 KiB API body ceiling, including other settings.
    if len(json.dumps([p.model_dump() for p in profiles], ensure_ascii=False).encode('utf-8')) > 48000:
        raise ValueError('School settings exceed 48 KB. Shorten notes or remove old terms.')


def public_school(profiles: list[SchoolProfile], members) -> list[dict]:
    """Mask at the server boundary, before display/offline JSON is delivered.

    Uniform and snack counts are operational status, like the rota. Descriptive
    titles, rooms and packing/pickup notes follow the person's title masking.
    """
    masked = {m.key for m in members if m.maskTitles}
    result = [p.model_dump() for p in profiles]
    for profile in result:
        if profile['member'] not in masked:
            continue
        for day in profile['days']:
            day['kit'] = ''
            for lesson in day['lessons']:
                lesson.update(name='Lesson', room='')
        for activity in profile['activities']:
            activity.update(name='Activity', room='', notes='')
    return result

File: web/school.mjs
/** School rules are civil-date patterns, not 24-hour recurrences or Google events. */
import {ordinal, isoDay} from './dates.mjs';
export const WEEKDAYS = ['Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday'];
export const UNIFORMS = {unknown:'Uniform not set',school:'School uniform',pe:'PE uniform',none:'No uniform'};
export const mod = (n, d) => ((n % d) + d) % d;
export const weekday = day => mod(day + 3, 7) + 1;
export const mondayOf = day => day - weekday(day) + 1;
const fail = message => { throw new Error(`School: ${message}`); };
function object(raw, allowed, name) {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) fail(`${name} must be an object.`);
  if (Object.keys(raw).some(k => !allowed.includes(k))) fail(`${name} contains an unknown field.`);
}
function text(value, max, name, required=false) {
  if (typeof value !== 'string' || value.length > max || /[\x00-\x1f\x7f]/.test(value) || (required && !value.trim())) fail(`Invalid ${name}.`);
  return value.trim();
}
function integer(value, min, max, name) {
  if (!Number.isInteger(value) || value < min || value > max) fail(`Invalid ${name}.`);
  return value;
}
function array(value, max, name) {
  if (!Array.isArray(value) || value.length > max) fail(`Too many or invalid ${name}.`);
  return value;
}
function date(value, anchor=false) {
  ordinal(value);
  if (anchor && weekday(ordinal(value)) !== 1) fail('Week A must be anchored to a Monday.');
  return value;
}
function dates(values, max) {
  const result=array(values ?? [],max,'exception dates').map(v=>date(v));
  if (new Set(result).size !== result.length) fail('Exception dates must be unique.');
  return result;
}
function lesson(raw, activity=false) {
  object(raw,['name','start','end','room',...(activity?['weekday','week','notes','skipDates']:[])],'lesson/activity');
  const result={name:text(raw.name,80,'lesson/activity name',true),start:raw.start??'',end:raw.end??'',room:text(raw.room??'',40,'room')};
  if (Boolean(result.start)!==Boolean(result.end)) fail('Enter both start and end times, or neither.');
  for (const t of [result.start,result.end]) if (typeof t!=='string'||(t&&!/^(?:[01]\d|2[0-3]):[0-5]\d$/.test(t))) fail('Use times in HH:MM format.');
  if (result.start && result.end <= result.start) fail('End time must be later on the same day.');
  if (activity) {
    if (!result.start) fail('Clubs and music lessons need start and end times.');
    Object.assign(result,{weekday:integer(raw.weekday,1,7,'activity weekday'),week:integer(raw.week??0,0,2,'activity week'),notes:text(raw.notes??'',160,'activity notes'),skipDates:dates(raw.skipDates,80)});
  }
  return result;
}
export function validateSchool(raw=[], members=[]) {
  const profiles=array(raw,6,'school profiles').map(p=>{
    object(p,['member','cycleWeeks','anchor','terms','excludeDates','days','activities'],'profile');
    if (!members.some(m=>m.key===p.member)) fail('Choose an existing person for each profile.');
    const cycleWeeks=integer(p.cycleWeeks??1,1,2,'cycle length');
    const terms=array(p.terms??[],32,'terms').map(t=>{
      object(t,['start','end','weekA'],'term');
      const term={start:date(t.start),end:date(t.end),weekA:t.weekA?date(t.weekA,true):''};
      if (term.end<term.start) fail('Term end must not precede its start.');
      return term;
    });
    const sorted=[...terms].sort((a,b)=>a.start.localeCompare(b.start));
    if (sorted.some((t,i)=>i && sorted[i-1].end>=t.start)) fail('Term ranges must not overlap; ends are inclusive.');
    const days=array(p.days??[],10,'school-day rules').map(d=>{
      object(d,['week','weekday','uniform','snacks','kit','lessons'],'day rule');
      const uniform=d.uniform??'unknown';
      if (!Object.hasOwn(UNIFORMS,uniform)) fail('Choose a supported uniform.');
      return {week:integer(d.week,1,cycleWeeks,'day week'),weekday:integer(d.weekday,1,5,'school weekday'),uniform,
        snacks:d.snacks==null?null:integer(d.snacks,0,9,'snack count'),kit:text(d.kit??'',120,'kit'),lessons:array(d.lessons??[],10,'lessons').map(l=>lesson(l))};
    });
    if (new Set(days.map(d=>`${d.week}:${d.weekday}`)).size!==days.length) fail('Duplicate school-day rules.');
    const activities=array(p.activities??[],24,'activities').map(a=>lesson(a,true));
    if (activities.some(a=>a.week>cycleWeeks)) fail('Week B activities require a two-week timetable.');
    return {member:p.member,cycleWeeks,anchor:date(p.anchor,true),terms,excludeDates:dates(p.excludeDates,180),days,activities};
  });
  if (new Set(profiles.map(p=>p.member)).size!==profiles.length) fail('Only one profile per person.');
  if (new TextEncoder().encode(JSON.stringify(profiles)).length>48000) fail('School settings exceed 48 KB. Shorten notes or remove old terms.');
  return profiles;
}
export function schoolDay(profile, day) {
  const iso=isoDay(day), term=profile.terms.find(t=>t.start<=iso && iso<=t.end);
  const base={member:profile.member,day,school:false,week:null,uniform:'unknown',snacks:null,kit:'',lessons:[],activities:[]};
  if (!profile.terms.length) return {...base,reason:'Term dates not set'};
  if (!term) return {...base,reason:'Outside saved term dates'};
  if (profile.excludeDates.includes(iso)) return {...base,reason:'No school / excluded date'};
  const week=mod(Math.floor((mondayOf(day)-ordinal(term.weekA||profile.anchor))/7),profile.cycleWeeks)+1;
  const dow=weekday(day), rule=profile.days.find(r=>r.week===week && r.weekday===dow);
  const activities=profile.activities.filter(a=>a.weekday===dow && (!a.week || a.week===week) && !a.skipDates.includes(iso)).sort((a,b)=>a.start.localeCompare(b.start));
  return {...base,...rule,day,week,school:dow<=5,reason:dow>5?'Weekend':'',activities};
}
export function snackLabel(count) { return count==null?'Snacks not set':count===0?'No snacks':`${count} snack${count===1?'':'s'}`; }
